src/dir.rs

```diff
@@ -0,0 +1,167 @@
+use {Error, Result};
+use errno::Errno;
+use libc;
+use std::os::unix::io::{AsRawFd, IntoRawFd, RawFd};
+use std::{ffi, fmt, ptr};
+
+#[cfg(target_os = "linux")]
+use libc::{dirent64 as dirent, readdir64_r as readdir_r};
+
+#[cfg(not(target_os = "linux"))]
+use libc::{dirent, readdir_r};
+
+/// An open directory.
+///! This is a lower-level interface than `std::fs::ReadDir`. Notable differences:
+///!    * can be opened from a file descriptor (as returned by `openat`, perhaps before knowing
+///!      if the path represents a file or directory).
+///!    * implements `AsRawFd`, so it can be passed to `fstat`, `openat`, etc.
+///!    * can be iterated through multiple times without closing and reopening the file
+///!      descriptor. Each iteration rewinds when finished.
+///!    * returns entries for `.` (current directory) and `..` (parent directory).
+///!    * returns entries' names as a `CStr` (no allocation or conversion beyond whatever libc
+///!      does).
+pub struct Dir(ptr::NonNull<libc::DIR>);
```

Comment:
This didn't go over well with the continuous build:

```
error[E0412]: cannot find type `NonNull` in module `ptr`
  --> src/dir.rs:23:21
   |
23 | pub struct Dir(ptr::NonNull<libc::DIR>);
   |                     ^^^^^^^ not found in `ptr`
```

so I switched to `*mut libc::DIR` instead.

Reply:
That's because `NonNull` only got added in Rust 1.25.0.  Nix's minimum requirement is 1.20.0.